### PR TITLE
[Darwin] MTRDeviceController getSessionForNode: to use subscription pool as needed

### PR DESCRIPTION
**Pull request by jtung-apple**
This change makes MTRDeviceController's `-getSessionForNode:completion:` also participate in the subscription pool when there is an MTRDevice associated with the nodeID, and is also known to use Thread.

With this change, MTRBaseDevice and MTRBaseClusters reads/subscriptions for devices using Thread would also get the CASE session establishment gated by the subscription pool, and help avoid congesting the Thread network when a controller starts and tries to get status for all its Thread devices.

### CODE REVIEW

File: src/darwin/Framework/CHIPTests/MTRPerControllerStorageTests.m [hunk 1]
```diff
@@ -2269,6 +2269,7 @@ - (void)doTestSubscriptionPoolWithSize:(NSInteger)subscriptionPoolSize deviceOnb
     XCTAssertNotNil(factory);
 
     __auto_type queue = dispatch_get_main_queue();
+    //    __auto_type queue = dispatch_queue_create("test.subscription.pool", DISPATCH_QUEUE_SERIAL_WITH_AUTORELEASE_POOL);
```
Comment (woody-apple) [inline src/darwin/Framework/CHIPTests/MTRPerControllerStorageTests.m:2272]:
```suggestion
```

File: src/darwin/Framework/CHIP/MTRDevice_Internal.h [hunk 2]
```diff
@@ -117,6 +117,9 @@ MTR_TESTABLE
 
 - (void)setStorageBehaviorConfiguration:(MTRDeviceStorageBehaviorConfiguration *)storageBehaviorConfiguration;
 
+// Returns whether this MTRDevice uses Thread for communication
+- (BOOL)deviceUsesThread;
```
Comment (woody-apple) [inline src/darwin/Framework/CHIP/MTRDevice_Internal.h:121]:
```suggestion
- (BOOL)_deviceUsesThread;
```

Comment (woody-apple) [inline src/darwin/Framework/CHIP/MTRDevice_Internal.h:121]: We should probably have a public or private bitmask for what type of transports it supports...

Comment (woody-apple) [reply src/darwin/Framework/CHIP/MTRDevice_Internal.h:121]: (More a note, I might take a stab at this)

Comment (jtung-apple) [reply src/darwin/Framework/CHIP/MTRDevice_Internal.h:121]: Right, this is currently just used internally to signal when to pool these attempts. I think things can be named in a more generic way..... Naming is a hard problem and we definitely need to give it more thought.

Comment (woody-apple) [reply src/darwin/Framework/CHIP/MTRDevice_Internal.h:121]: It's fine for now :)

**Review by woody-apple — APPROVED** [review]

File: src/darwin/Framework/CHIP/MTRDevice.mm [hunk 3]
```diff
@@ -2055,194 +2061,195 @@ - (void)_setupSubscriptionWithReason:(NSString *)reason
         });
     }
 
+    // Call directlyGetSessionForNode because the subscription setup already goes through the subscription pool queue
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/MTRDevice.mm:2064]: So I was figuring we would stop putting the subscription setup in there and just call getSessionForNode and have _that_ get queued.... why are we not doing that?

Comment (jtung-apple) [reply src/darwin/Framework/CHIP/MTRDevice.mm:2064]: Subscriptions on controller startup can cause a thundering herd issue with many Thread devices. If we only use the N-pool for getSessionForNode, then as soon as session is established, another device subscription can move forward, and we can get into a situation where many subscriptions and their priming reports are in the air at the same time.

**Pull request merged**